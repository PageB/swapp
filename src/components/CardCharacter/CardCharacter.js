import React from 'react';
import CardRowData from '../CardRowData/CardRowData';
import PropTypes from 'prop-types';

import styles from './CardCharacter.module.scss';

const cardCharacter = props => {
  return (
    <div className={[styles.Card, styles[props.theme]].join(' ')}>
      <p className={styles.Card__Title}>{props.card.name}</p>
      <img className={styles.Card__Image} src={props.card.image} alt={props.card.name} />
      <div className={styles.Card__TextDetails}>
        <CardRowData label="Height: ">{props.card.height}</CardRowData>
        <CardRowData label="Weight: ">{props.card.mass}</CardRowData>
        <CardRowData label="Species: ">{props.card.species}</CardRowData>
        <CardRowData label="Home World: ">{props.card.homeworld}</CardRowData>
      </div>
    </div>
  );
};

cardCharacter.propTypes = {
  theme: PropTypes.oneOf(['DarkTheme', 'LightTheme']),
};

cardCharacter.defaultProps = {
  theme: 'DarkTheme',
};

export default cardCharacter;
