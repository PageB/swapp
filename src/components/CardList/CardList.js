import React, { Component } from 'react';
import PropTypes from 'prop-types';

import styles from './CardList.module.scss';

class cardList extends Component {
  /**
   * Loop all cards and render each as JSX item.
   *
   * @method renderCardItems
   * @param {JSX Object} CardItem
   * @param {Object} cards
   * @return {JSX Object}
   */
  renderCardItems(cards, CardItem, cardTheme, cardNavigation) {
    return cards.map(card => {
      return <CardItem key={card.id} card={card} theme={cardTheme} navigate={cardNavigation} />;
    });
  }

  /**
   * Render lifecycle hook
   *
   * @method render
   * @return {Object}
   */
  render() {
    const { cards, component: CardItem, theme: cardTheme, cardNavigation, direction } = this.props;

    return (
      <div className={styles.CardList} style={{ flexDirection: direction }}>
        {this.renderCardItems(cards, CardItem, cardTheme, cardNavigation)}
      </div>
    );
  }
}

cardList.propTypes = {
  cards: PropTypes.array,
  direction: PropTypes.oneOf(['column', 'row']),
  theme: PropTypes.oneOf(['DarkTheme', 'LightTheme']),
};

cardList.defaultProps = {
  direction: 'row',
  theme: 'DarkTheme',
};

export default cardList;
