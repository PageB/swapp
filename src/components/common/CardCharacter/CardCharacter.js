import React from 'react';
import styles from './CardCharacter.module.css';
import CardRowData from '../CardRowData/CardRowData';

const cardCharacter = props => {
  return (
    <div className={styles.Card}>
      <p className={styles.CardName}>{props.card.name}</p>
      <img className={styles.CardImage} src={props.card.image} alt={props.card.name} />
      <div className={styles.CardDetails}>
        <CardRowData label="Height: ">{props.card.height}</CardRowData>
        <CardRowData label="Weight: ">{props.card.weight}</CardRowData>
        <CardRowData label="Species: ">{props.card.species}</CardRowData>
        <CardRowData label="Home World: ">{props.card.world}</CardRowData>
      </div>
    </div>
  );
};

export default cardCharacter;
