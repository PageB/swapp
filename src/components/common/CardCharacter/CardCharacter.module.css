.Card {
  background-color: #333333;
  border: 1px solid #3c4858;
  border-radius: 8px;
  color: #ffe300;
  display: flex;
  flex-direction: column;
  width: 280px;
}

.CardName {
  border-radius: 8px 8px 0px 0px;
  color: inherit;
  font-weight: bold;
  font-size: 18px;
  text-align: center;
}

.CardImage {
  border-radius: 8px;
  padding: 0 16px;
  width: 248px;
}

.CardDetails {
  color: inherit;
  background-color: inherit;
  padding: 0 16px;
  font-size: 14px;
}
