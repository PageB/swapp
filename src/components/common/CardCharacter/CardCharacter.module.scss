@import '../../../colors.scss';

.Card {
  background-color: $dark-charcoal;
  border-radius: 8px;
  color: $vivid-yellow;
  display: flex;
  flex-direction: column;
  width: 280px;
}

.CardName {
  border-radius: 8px 8px 0px 0px;
  color: inherit;
  font-weight: bold;
  font-size: 18px;
  text-align: center;
}

.CardImage {
  border-radius: 8px;
  padding: 0 16px;
  width: 248px;
}

.CardDetails {
  color: inherit;
  background-color: inherit;
  padding: 0 16px;
  font-size: 14px;
  border-radius: 8px;
}

.DarkTheme {
  background-color: $dark-charcoal;
  color: $vivid-yellow;
}

.LightTheme {
  background-color: $white;
  color: $crayola;
}
