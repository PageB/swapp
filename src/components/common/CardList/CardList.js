import React, { Component } from 'react';
import styles from './CardList.module.css';

class cardList extends Component {
  /**
   * Loop all cards and render each as JSX item.
   *
   * @method renderCardItems
   * @param {JSX Object} CardItem
   * @param {Object} cards
   * @return {JSX Object}
   */
  renderCardItems(CardItem, cards) {
    return cards.map(card => {
      return <CardItem key={card.id} card={card}/>;
    });
  }

  /**
   * Render lifecycle hook
   *
   * @method render
   * @return {Object}
   */
  render() {
    const { component: CardItem, cards } = this.props;

    return <div className={styles.CardList}>{this.renderCardItems(CardItem, cards)}</div>;
  }
}

export default cardList;
