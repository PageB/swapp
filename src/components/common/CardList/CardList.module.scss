@import '../../../colors.scss';

.CardList {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  padding: 16px;
}
