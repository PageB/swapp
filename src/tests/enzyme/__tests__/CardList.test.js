import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { starWarsCharacters } from '../../../mock/data';

import CardCharacter from '../../../components/CardCharacter/CardCharacter';
import CardList from '../../../components/CardList/CardList';

configure({ adapter: new Adapter() });

describe('CardList Component', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(
      <CardList cards={starWarsCharacters} component={CardCharacter} direction="row" />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
