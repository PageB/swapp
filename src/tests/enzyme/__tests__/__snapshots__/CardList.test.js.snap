// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList Component should match snapshot 1`] = `
<div
  className="CardList"
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 1,
        "image": "https://upload.wikimedia.org/wikipedia/en/3/32/Ben_Kenobi.png",
        "mass": 77,
        "name": "Obi-Wan Kenobi",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="1"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 2,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/rs-218591-5-leia-a09cab78-e7a1-4be4-b437-99a1d7d4d280.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "Princess Leia",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="2"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 3,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/rs-218589-4-yoda-6a430ef2-4be5-43d5-8614-e743ca7b2f69.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "Yoda",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="3"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 4,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/rs-218587-3-boba-fett-9dce897b-f605-4fae-9de7-dfbf95ab1a2d.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "Boba Fett",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="4"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 5,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/rs-218586-2-vader-f6cdf838-912c-4bfa-8a12-f43521fbd99c.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "Darth Vader",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="5"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 6,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/rs-218584-1-han-f4c9a887-81df-4c58-b166-94ab4485286b.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "Han Solo",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="6"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 7,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/bb8.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "BB-8",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="7"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 8,
        "image": "https://www.rollingstone.com/wp-content/uploads/2018/06/rs-218664-7-r2d2-cc3a79ac-5212-4452-852d-27eacea91652.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "R2-D2",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="8"
    theme="DarkTheme"
  />
  <cardCharacter
    card={
      Object {
        "birthYear": "",
        "height": 182,
        "homeworld": "Stewjon",
        "id": 9,
        "image": "https://www.rollingstone.com/wp-content/uploads/2015/12/rs-218592-6-luke-cd70ba0f-828e-4231-9c8d-f8d6ab92a07b.jpg?resize=1240,1240&w=385",
        "mass": 77,
        "name": "Luke Skywalker",
        "species": "Human",
        "starships": Array [
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 100,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 200,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
          Object {
            "cost": 631,
            "crew": 108,
            "hyperdriveRating": 150,
            "id": 300,
            "image": "https://www.super-hobby.com/zdjecia/7/1/9/20828_rd.jpg",
            "maxAtmosphericSpeed": 1159,
            "maxMLPerHour": 631,
            "model": "Delta-7 Aethersprite-class Starfighter",
            "name": "Jedi Starfighter",
            "starshipClass": "starfighter",
          },
        ],
      }
    }
    key="9"
    theme="DarkTheme"
  />
</div>
`;
